pyschema/source_generation.py

```diff
@@ -22,11 +24,111 @@ def to_python_source(classes, indent=DEFAULT_INDENT):
+        return namespace
+
+    def get_namespace_clusters(self, all_classes):
+        namespace_cluster = defaultdict(set)
+        for c in all_classes:
+            namespace = self.get_namespace(c)
+            namespace_cluster[namespace].add(c)
+        return namespace_cluster
+
+    def format_definitions(self, classes):
+        return "\n\n".join([_class_source(c, self.indent) for c in classes])
+
+    def write_namespace_file(self, namespace, module_code):
+        if namespace is None:
+            key = ['__init__']
+        else:
+            key = namespace.split('.')
+        output_file = os.path.join(self.target_folder, *key) + '.py'
+        output_dir = os.path.dirname(output_file)
+        if not os.path.exists(output_dir):
+            os.makedirs(output_dir)
+
+        with open(output_file, 'w') as out_fn:
+            out_fn.write(module_code)
+
+    def write_init_files(self):
+        def touch_init_file_in_folder(folder):
+            path = os.path.join(folder, '__init__.py')
+            if not os.path.exists(path):
+                open(path, 'w').close()
+
+        touch_init_file_in_folder(self.target_folder)
+        for root, dirs, _ in os.walk(self.target_folder):
+            for d in dirs:
+                touch_init_file_in_folder(os.path.join(root, d))
+
+    def format_imports(self, imported_classes):
+        if not imported_classes:
+            return "\n"
+        imported_namespaces = self.get_namespace_clusters(imported_classes)
+        lines = []
+        for namespace, schemas in imported_namespaces.iteritems():
+            if self.parent_package:
+                if namespace:
+                    module = "{}.{}".format(self.parent_package, namespace)
+                else:
+                    module = self.parent_package
+            else:
+                if not namespace:
+                    module = "."
+                else:
+                    module = namespace
+            class_part = ", ".join(s.__name__ for s in schemas)
+            lines.append("from {} import {}".format(module, class_part))
+        return "\n".join(lines) + "\n\n"
+
+    def from_classes_with_refs(self, classes):
+        class_graph = CachedGraphTraverser()
+
+        all_classes = set(classes)
+        for c in classes:
+            referenced_schemas = class_graph.find_descendants(c)
+            all_classes |= set(referenced_schemas)
+
+        namespace_cluster = self.get_namespace_clusters(all_classes)
+
+        ordered_schemas = class_graph.get_reference_ordered_schemas(all_classes)
+
+        for namespace, classes in namespace_cluster.iteritems():
+            inlined_classes = [c for c in ordered_schemas if c in classes]
+            imported_classes = set()
+
+            for inlined in inlined_classes:
+                all_referenced = class_graph.find_descendants(inlined)
+                imported_classes |= set([c for c in all_referenced if c not in inlined_classes])
```

Comment:
We are adding un-needed imports here. Not a biggie to have unused imports, but only children/direct descendant are needed here, since all others will only be referenced in the imported files where they in turn are imported... Should be a small modification to fix this

Test case checking for this (and currently failing) is part of PR